spec!: add support for several topic matcher types, including URL Pattern

dunglas opened the pull request:
Replaces https://github.com/dunglas/mercure/pull/831 with a more flexible and future-proof approach.

Closes #588.

## Summary

* In addition to URI templates, [regular expressions](https://datatracker.ietf.org/doc/rfc9485/), [URL patterns](https://urlpattern.spec.whatwg.org/) and [CEL](https://cel.dev/) expressions are now supported
* the `topic` query parameter becomes `match` and now only supports exact matching
* new query parameters `match{MatchType}` are introduced
  * by default, `matchRegexp`, `matchURLPattern`, `matchCEL`, and `matchURITemplate` are defined
  * hub can implement custom matcher types
  * future updates may introduce new matcher types
* it's now possible to include per-topic (or topic matcher) payloads (replaces #945)
* the JWT format has been updated to match these new capabilities
* the subscription API format has been updated to match these new capabilities

The official hub will be able to support both the current and new protocols simultaneously (no hard BC break).

TODO:

* [x] update all examples

dunglas committed:
spec!: add support for several topic matcher types, including URL Pattern

soyuka commented on spec/mercure.md:
```diff
@@ -30,7 +30,8 @@ organization = "Les-Tilleuls.coop"
 a fast, reliable and battery-efficient way. It is especially useful for publishing real-time updates
-of resources served through sites and web APIs to web and mobile apps.
+of resources served through sites and web APIs to web and mobile apps, but can also be used
+as a simple publish subcribe system.
```
```suggestion
as a simple publish subscribe system.
```

soyuka commented on spec/mercure.md:
```diff
@@ -44,165 +45,27 @@ The keywords **MUST**, **MUST NOT**, **REQUIRED**, **SHALL**, **SHALL NOT**, **S
+*   Topic matcher: An expression intended to be matched by one or several topics,
+    depending on the matcher type.
+*   Topic matcher type: Several types of matchers **MAY** be supported by the hub. The hub **MUST** support exact matching of topic and **SHOULD** support
+    matching topics using I-Regexp [@!RFC9485], URL patterns [@!urlpattern] and URI Templates [@!RFC6570] as
```
I'd say: 

should support at least one of the following topic matchers:

dunglas replied:
I'll improve this by adding a table.

soyuka commented on spec/mercure.md:
```diff
@@ -785,14 +687,142 @@ The JSON-LD context available at `https://mercure.rocks` is the following:
+Link: <https://example.com/.well-known/mercure>; rel="mercure"
+
+{"@id": "/books/foo", "foo": "bar", "@context": {"@language": "fr-FR"}}
+~~~
```
nothing has changed there it just got moved?

dunglas replied:
yes

soyuka commented on spec/mercure.md:
```diff
@@ -213,20 +76,29 @@ superior to leverage multiplexing and other performance-oriented related feature
+The subscriber receives updates for all topics matching at least a topic matcher according to
+the matcher type rules.
+
+The hub **SHOULD** support the `Regexp`, `URLPattern` and `URITemplate` matcher types.
```
It should support all of them or only at least one?

dunglas replied:
All of them. But I'll change this to have regexp and URL Pattern as "should" and URI Template" as "may".

soyuka commented on spec/mercure.md:
```diff
@@ -213,20 +76,29 @@ superior to leverage multiplexing and other performance-oriented related feature
+the matcher type rules.
+
+The hub **SHOULD** support the `Regexp`, `URLPattern` and `URITemplate` matcher types.
+The corresponding query parameters are respectively `matchRegexp`, `matchURLPattern` and `matchURITemplate`.
```
```suggestion
The corresponding query parameters are respectively `matchRegExp`, `matchURLPattern` and `matchURITemplate`.
```

dunglas replied:
The RFC use the `Regexp` case, let's be consistent with it: https://www.ietf.org/rfc/rfc9485.html

soyuka commented on spec/mercure.md:
```diff
@@ -240,11 +112,11 @@ partial update by using formats such as JSON Patch [@RFC6902] or JSON Merge Patc
 supported by hubs.
 
-The resource **SHOULD** be represented in a format with hypermedia capabilities such as
+The resource **MAY** be represented in a format with hypermedia capabilities such as
```
Why the change?

dunglas replied:
I want to relax the hypermedia part. With the introduction of the regexp matcher, hypermedia just becomes a use case among others.

soyuka commented on spec/mercure.md:
```diff
@@ -258,7 +130,7 @@ Example:
 url.searchParams.append('topic', 'https://example.com/foo');
 url.searchParams.append('topic', 'bar');
-url.searchParams.append('topic', 'https://example.com/bar/{id}');
+url.searchParams.append('matchurlpattern', 'https://example.com/bar/:id');
```
```suggestion
url.searchParams.append('matchURLPattern', 'https://example.com/bar/:id');
```

soyuka commented on spec/mercure.md:
```diff
@@ -240,11 +112,11 @@ partial update by using formats such as JSON Patch [@RFC6902] or JSON Merge Patc
 [@W3C.REC-html52-20171214].
 
-Web Linking [@!RFC5988] **SHOULD** be used to indicate the IRI of the resource sent in the event.
+Web Linking [@!RFC5988] **MAY** be used to indicate the IRI of the resource sent in the event.
```
why the change?

dunglas replied:
same

soyuka commented on spec/mercure.md:
```diff
@@ -411,12 +283,17 @@ authorized to send updates for the specified topics.
+array of topic matchers. See (#subscription).
+
+Topic matchers must **MUST** be strings, in which case they will **MUST** be matched exactly by the topic,
+or objects with the a `match` entry containing the topic matcher itself, and an optional `matchType` entry containing the topic matcher
```
I did not understood the sentence `will **MUST**`, `with the a match`

dunglas replied:
clarified/fixed

soyuka commented on spec/mercure.md:
```diff
@@ -411,12 +283,17 @@ authorized to send updates for the specified topics.
+
+Topic matchers must **MUST** be strings, in which case they will **MUST** be matched exactly by the topic,
+or objects with the a `match` entry containing the topic matcher itself, and an optional `matchType` entry containing the topic matcher
+type. If no `matchType` key is present, exact matching **MUST** ne used. The hub **SHOULD** support the
```
```suggestion
type. If no `matchType` key is present, exact matching **MUST** be used. The hub **SHOULD** support the
```

soyuka commented on spec/mercure.md:
```diff
@@ -411,12 +283,17 @@ authorized to send updates for the specified topics.
+Topic matchers must **MUST** be strings, in which case they will **MUST** be matched exactly by the topic,
+or objects with the a `match` entry containing the topic matcher itself, and an optional `matchType` entry containing the topic matcher
+type. If no `matchType` key is present, exact matching **MUST** ne used. The hub **SHOULD** support the
+`Regexp`, `URLPattern` and `URITemplate` topic matcher types and **MAY** support other types.
```
```suggestion
`RegExp`, `URLPattern` and `URITemplate` topic matcher types and **MAY** support other types.
```

soyuka commented on spec/mercure.md:
```diff
@@ -479,27 +356,43 @@ and in subscription events.
+        },
+        {
+            "match": "https://example.com/bar/:id",
+            "matchType": "urlpattern",
```
now the case is not important? why not CamelCase like the matchtypes used inside query parameters?

dunglas replied:
fixed

soyuka commented on spec/mercure.md:
```diff
@@ -592,11 +490,12 @@ least the following properties:
     topic
 *   `type`: the fixed value `Subscription`
-*   `topic`: the topic selector used of this subscription
+*   `matchType`: the topic matcher type used of this subscription (**MUST** be omitted if the matcher type is exact matching)
```
about what I said earlier, we should specify whether this is camel case, not case sensitive or lowercase

dunglas replied:
done

dunglas committed:
Update spec/mercure.md

Co-authored-by: Antoine Bluchet <soyuka@users.noreply.github.com>

dunglas committed:
Update spec/mercure.md

Co-authored-by: Antoine Bluchet <soyuka@users.noreply.github.com>

dunglas committed:
Update spec/mercure.md

Co-authored-by: Antoine Bluchet <soyuka@users.noreply.github.com>

dunglas committed:
Update mercure.md

dunglas committed:
better matcher type handling

soyuka commented on spec/mercure.md:
```diff
@@ -271,6 +146,58 @@ eventSource.onmessage = function ({data}) {
+
+The hub **MAY** support using CEL expressions [@cel] as matchers.
+
+A variable named `topics` containing an array of string **MSUT** be passed
```
```suggestion
A variable named `topics` containing an array of string **MUST** be passed
```

soyuka commented on spec/mercure.md:
```diff
@@ -271,6 +146,58 @@ eventSource.onmessage = function ({data}) {
+The expression **MUST** return a boolean value: true if the topic matches, false otherwise.
+
+If parsing or checking of a CEL expression fails or if the expression does not return a boolean value, the hub **MUST** return a 400 "Bad Request" HTTP status code.
+
```
using topics replaces matcher types? Or you can use both topics and matchers together? Maybe that we're missing some precision about this.

dunglas replied:
I mean, you will have access to a variable named `topics` to use in the CEL expression itself. The matcher type is CEL.

soyuka replied:
The matcher type name is `URITemplate`.
The corresponding query parameter is `matchURITemplate`.

dunglas committed:
Update spec/mercure.md

Co-authored-by: Antoine Bluchet <soyuka@users.noreply.github.com>

soyuka commented on spec/mercure.md:
```diff
@@ -213,39 +75,52 @@ superior to leverage multiplexing and other performance-oriented related feature
+In addition to `match` query parameters, the subscriber can pass other topic matchers by passing
+query parameters starting with the string `match` and followed by the topic matcher type.
+
+The `matcExact` query parameter **MUST** be considered the same as `match`.
```
```suggestion
The `matchExact` query parameter **MUST** be considered the same as `match`.
```

dunglas committed:
add matcher name and query parameter for CEL

dunglas committed:
Update spec/mercure.md

Co-authored-by: Antoine Bluchet <soyuka@users.noreply.github.com>

soyuka commented on spec/mercure.md:
```diff
@@ -213,39 +75,52 @@ superior to leverage multiplexing and other performance-oriented related feature
-interface](https://html.spec.whatwg.org/multipage/server-sent-events.html#the-eventsource-interface)
-**MAY** be used to establish the connection. Any other appropriate mechanism
+The protocol doesn't specify the maximum number of query parameters that can be sent, but the hub
+**MAY** apply an arbitrary limit. A subscription is created for every provided parameter starting with the string `match`. See (#subscription-events).
```
```suggestion
**MAY** apply an arbitrary limit. A subscription is created for every provided parameter starting 
with the string `match`. See (#subscription-events).
```

dunglas committed:
Update spec/mercure.md

Co-authored-by: Antoine Bluchet <soyuka@users.noreply.github.com>

soyuka commented:
This is still ongoing work

dunglas committed:
review

dunglas committed:
Various improvements

dunglas committed:
review

dunglas commented on spec/mercure.md:
```diff
@@ -431,136 +373,192 @@ If the presented JWS contains an expiration time in the standard `exp` claim def
-For instance, `mercure.payload` can contain the user ID of the subscriber, a list of groups it
-belongs to, or its IP address. Storing data in `mercure.payload` is a convenient way to share data
-related to one subscriber to other subscribers.
+Each entry in the `mercure.subscribe` claim of the JWS **MAY** contain a JSON object under the `payload` key.
```
Clarified - the spec now states that entries MUST either be strings or objects, and payloads are only applicable to object-form entries.

dunglas committed:
spec: clarify payload matching, URL Pattern base URL, and CEL DoS handling

Three clarifications surfaced while implementing this spec revision:

- Payload assignment: define what it means for a JWT claim matcher to
  "match" a subscription matcher. A claim matches when it is the
  wildcard `*`, when its (type, pattern) pair equals the subscription's,
  or when its matcher accepts the subscription's `match` string as a
  topic. Payload fallback to top-level `mercure.payload` is stated
  explicitly.

- URL Pattern base URL: recommend absolute patterns and describe the
  hub's freedom in resolving relative ones. Subscribers and publishers
  should not rely on the base URL being anything specific because
  resolution is implementation-defined.

- CEL evaluation cost limit: hubs SHOULD enforce an implementation-
  defined cost limit to mitigate DoS from pathological expressions in
  hostile JWTs; exceeding the limit yields `false`.

Also notes that hubs MAY reject oversized patterns with a 400 response.

dunglas committed:
spec: require object-form JWT matcher claims outside compat mode

Silently treating v8 bare-string claims as `Exact` is an interoperability
trap: the v8 rule was "exact OR URI Template", so a token signed for a
v8 hub would match fewer topics on a v9 hub without any diagnostic.

Bare-string claims are now only accepted when the hub is explicitly
running in protocol-compatibility mode; v9 hubs reject them with 401.
This forces clients to migrate to the unambiguous object form.

dunglas committed:
spec: require URL patterns to be absolute

Topics are absolute IRIs by design, and the protocol has no notion of a
base URL against which a subscriber could expect a relative pattern to
resolve. Allowing relative patterns forced hubs to invent an implicit
base (typically the hub's own location), which made subscriptions
non-portable across hubs and the resolution implementation-defined.

Hubs MUST now reject relative URL patterns. The Go implementation in
PR #1207 already does this.

dunglas committed:
spec: editorial pass for clarity and IETF style

dunglas committed:
spec: drop implementation-specific compatibility-mode language

dunglas committed:
spec: allow relative URL patterns, resolved against the hub URL

The hub publishes subscription events on relative topic IDs
(e.g., /.well-known/mercure/subscriptions/...). Requiring patterns
to be absolute prevents subscribers from matching those topics
with matchURLPattern.

dunglas committed:
spec: editorial pass on topic matchers, case-sensitive params, IANA registry

- Bump draft to -08
- Drop redundant "exactly one of" wording in Subscription
- Add (#discovery) cross-reference from Subscription
- Switch topic matcher query parameter names to case-sensitive,
  registered in a new IANA "Mercure Topic Matcher Query Parameter Names"
  registry
- URL Pattern: MUST NOT enable ignoreCase; host case-insensitivity
  retained per URL canonicalization
- Exact and Regexp: MUST NOT resolve relative values against a base URL
- Topic Matcher List: reject non-object entries with 400 (was 401 bare-string),
  matchType value is case-sensitive and matches the canonical name from
  the matcher-type table
- Payloads: define top-level mercure.payload, drop the redundant "*"
  wildcard rule (catchall expressible via regex .*)
- Subscription Events: {matchType} URL component uses canonical case;
  JSON-LD matchType field case-sensitive
- Fix dart_mercure underscore typo

Signed-off-by: Kévin Dunglas <kevin@dunglas.fr>

dunglas committed:
spec: comprehensive security pass

Address security issues identified in a section-by-section audit. Adds
normative requirements to close authentication, authorization, denial-
of-service, injection, and information-disclosure gaps that were
previously implicit or unspecified.

Terminology and Subscription
- Constrain topic strings to UTF-8 and forbid C0 controls and DEL
- Bump per-request matcher count and pattern length limits from MAY
  to SHOULD, with 400 on overflow
- Require UTF-8 and control-char rejection for matcher query values
- Reword the subscription creation rule to match the registered query
  parameter set; allow hubs to deduplicate identical matchers

Matcher Types
- Exact: forbid relative resolution and implicit normalization; add
  guidance on NFC and IDNA canonicalization
- URL Pattern: require an RE2-class engine or evaluation cost/time
  limit; warn about scheme-wildcard hazards (data:, javascript:, etc.)
- Regular Expression: require an RE2-class engine or evaluation
  cost/time limit; I-Regexp is a dialect, not an engine
- CEL: bar custom functions from network, filesystem, process, or
  other side-effectful access; bump evaluation cost limit from SHOULD
  to MUST; cap topics array size
- Other Matcher Types: register names in the new IANA "Mercure
  Matcher Types" registry; vendor-prefix implementation-specific names

Publication
- Reserve /.well-known/mercure/ for hub-generated topics; reject
  publish requests violating the rule with 403
- Make the private field's truthiness rule explicit (presence wins,
  value ignored), preventing accidental opt-out interpretations
- Forbid C0 controls in id, LF/CR in type, and require ASCII digits
  for retry, with 400 on violation
- Require UTF-8 for field names, values, and data
- Cap request body and field length; reject overflow with 413

Authorization
- Add a JWS Validation subsection mandating rejection of alg: none,
  alg/key-type binding, signature verification, exp/nbf enforcement,
  and optional aud validation; require minimum alg support of EdDSA,
  ES256, and RS256
- Distinguish 401 (JWS missing/invalid) from 403 (insufficient scope);
  forbid disclosing which one failed
- Promote different-key-per-role guidance from MAY to SHOULD
- Promote cookie Secure and HttpOnly from SHOULD to MUST; default
  SameSite=Strict, Lax permitted only for cross-site discovery flows
- Topic Matcher List: any per-entry parse failure rejects the whole
  request with 400 (no partial acceptance, which could alter scope);
  cap matcher count and pattern length, with 400 on overflow
- Payloads: warn that first-match-wins ordering can mask specific
  entries; warn that payloads broadcast to other authorized
  subscribers via subscription events and must not carry per-
  subscriber identifiers

Reconnection
- Require authorization re-check on replayed events against the
  current JWS scope (not the scope at publication time)
- Forbid the Last-Event-ID response header from disclosing the
  identifier of an event the subscriber is not authorized to receive

Active Subscriptions
- Subscriber identifier is hub-assigned; clients MUST NOT supply,
  suggest, or override it; hub MUST guarantee uniqueness and MAY
  derive the value from the JWS sub claim
- Recommend correlating subscriptions of the same client across
  requests (e.g., by sub claim)

JSON-LD Context
- Subscribers SHOULD NOT auto-fetch the @context URL; embed or cache
  it locally to prevent MITM-induced semantic drift

Discovery
- Warn that a compromised publisher can redirect subscribers to a
  malicious hub; recommend restricting hub origins
- Tighten key-set hosting language; require publisher-side access
  control on the JWK Set endpoint; warn that misconfiguration defeats
  encryption

Encryption
- Restrict JWE algorithms to currently strong choices; forbid
  algorithms with known compromises (e.g., RSA1_5)
- Note that JWE has no replay protection; recommend in-payload
  freshness indicators
- Note that long-lived JWE keys lack forward secrecy; recommend
  rotation and ephemeral key agreement

IANA
- Add the "Mercure Matcher Types" registry (Specification Required)
  with initial entries Exact, URLPattern, Regexp, CEL, URITemplate
- Update query-parameter registry to reference the matcher-types
  registry

Security Considerations
- Add subsections summarizing the new threats: JWS validation, SSE
  field injection, reserved hub namespace, replay authorization,
  subscriber identifier assignment, regex/URL Pattern DoS, CEL
  sandbox, payload privacy, topic normalization, resource limits,
  hub trust, JWE algorithms and replay

References
- Add UNICODE and re2 bibliography entries

Signed-off-by: Kévin Dunglas <kevin@dunglas.fr>

dunglas committed:
spec: review fixes — RFC 2119 keyword, per-token limit, IANA change controller

- Replace ambiguous "No topic value MAY start with the prefix
  /.well-known/mercure/" with "Topic values MUST NOT start with the
  prefix". MAY is a permission keyword; combining it with "No" mixed
  RFC 2119 semantics with English negation.
- Add a normative body rule for per-JWS concurrent subscription
  limits (MAY + 429), aligning the body with the Resource Limits
  entry in Security Considerations.
- State "The change controller for all initial entries is the IETF"
  in both new IANA registries (Topic Matcher Query Parameter Names
  and Matcher Types), matching the convention used by the other
  IANA registrations in this document.

Signed-off-by: Kévin Dunglas <kevin@dunglas.fr>

dunglas committed:
spec: merge IANA query parameter and matcher type registries

Collapse the two newly added IANA registries into one. Query parameter
names are mechanically derived from matcher type names (match +
TypeName, plus the historical match alias for Exact), so maintaining
two separate registries adds editorial weight without adding
information. The merged registry binds the type name and its query
parameter at the moment of registration, making the relationship
explicit and reducing the chance of drift.

Signed-off-by: Kévin Dunglas <kevin@dunglas.fr>

dunglas committed:
spec: tighten rules for non-registered (hub-specific) matcher types

Custom matcher types extend the protocol surface and the attack
surface. Without explicit constraints, hubs implementing them risk
authorization bypass through type confusion, denial of service via
unguarded pattern evaluation, and SSRF or sensitive-material
disclosure via expression-based matchers.

- Require non-registered names to use reverse-DNS dotted notation
  (com.example.Foo) or an absolute HTTPS URI rooted in a
  domain controlled by the implementer. The PascalCase production
  is reserved for registered names; collisions are syntactically
  prevented.
- Require parity with built-in DoS and sandbox controls: linear-time
  matching engine or evaluation cost limit, no network/filesystem/
  process/clock/random access from expressions.
- Add an authorization invariant: a custom matcher MUST NOT produce
  an authorization decision unreachable by some registered matcher
  type for the same match value and topic.
- State explicitly that JWSs using non-registered names are not
  portable, and that incompatible hubs reject with 501.
- Add a Security Considerations subsection summarizing the rationale
  and requiring hub operators to review each custom matcher before
  enabling it.

Signed-off-by: Kévin Dunglas <kevin@dunglas.fr>

dunglas committed:
spec: refine custom matchers, JWS validation, replay, subscriber ID

Refinements following further review.

- Custom matchers
  - Drop the URI-form name option (matchhttps%3A%2F%2F... was
    parser-hostile after query-key percent-encoding).
  - Relax to "vendor-controlled prefix separated by .": DNS-reverse
    domain (com.example.Foo) or trademark (ACME.Foo). The intent is
    explicitly the inverse of RFC 6648's deprecated X- convention —
    require globally-unique prefixes rather than ambiguous markers.
  - Update the Subscription query-parameter rule so hub-supported
    non-registered names are accepted, not only IANA-registered ones.

- Authorization
  - Note that hubs MAY be deployed without JWS-based authorization
    when the network is fully trusted and the hub is unreachable from
    untrusted clients. Avoids relaxing alg=none, which would be
    unsafe even on intranets.
  - Compress JWS Validation by delegating to RFC 8725 (JWT BCP).
    Keep the high-impact items (alg=none, alg/key binding, exp/nbf,
    aud, recommended alg list) inline; defer the rest.

- Reconnection
  - Replace the predecessor-event lookup with a two-state Last-Event-
    ID response: same identifier if the requested event existed and
    was visible to the subscriber, earliest otherwise. Avoids both
    the backward-scan DoS vector and the leak of unauthorized event
    identifiers.

- Subscription Events
  - Subscriber identifier: drop the strict "hub-generated" rule.
    Reality: the JWS is minted by the application server, not the
    hub. Reframe as a forgery-prevention rule: clients MUST NOT
    forge, supply, or override the identifier; the hub MUST derive
    it from cryptographically-validated information (typically the
    sub claim). Updated Security Considerations subsection
    correspondingly.

Signed-off-by: Kévin Dunglas <kevin@dunglas.fr>

dunglas committed:
spec: Last-Event-ID — bounded backward scan with auth filter

Switch Last-Event-ID response semantics from the binary two-state
form back to a precise predecessor identifier, with two guards:

- the predecessor MUST be visible to the requesting subscriber
  (private events the subscriber cannot read are never disclosed
  via this header);
- the lookup is bounded by an implementation-defined backward-scan
  limit, preventing it from becoming a DoS vector through ancient
  Last-Event-ID values.

When the cap is hit or no visible predecessor is found, the hub
returns "earliest". This restores the event-store use case
(precise recovery anchor) while keeping the security and DoS
protections of the previous version.

Signed-off-by: Kévin Dunglas <kevin@dunglas.fr>

dunglas committed:
spec: relax Last-Event-ID scan limit and clarify ID-type leakage

The previous wording required hubs to enforce a backward-scan limit
unconditionally. The real cost of the lookup is per-event
authorization filtering, not the predecessor seek itself, which is
O(log N) on backends such as Redis Streams (XREVRANGE) and constant
on b-tree indexes.

- Demote the scan-limit rule from MUST to SHOULD, and explicitly
  permit hubs with cheap predecessor seek and cheap per-event
  authorization filtering to omit it entirely.
- Keep the authorization-filter constraint on the response header
  unchanged (MUST NOT disclose unauthorized event identifiers).
- Add an operator note explaining that the privacy impact of
  predecessor disclosure depends on the hub's identifier format:
  opaque random IDs (UUIDv4) leak only existence, while time-
  ordered IDs (UUIDv7, Redis Stream IDs, snowflake) additionally
  leak timing and ordering of events the subscriber cannot read.
  Operators handling highly sensitive private updates SHOULD prefer
  opaque random identifiers.

Signed-off-by: Kévin Dunglas <kevin@dunglas.fr>

dunglas committed:
spec: IETF-style review pass across protocol, security, refs, metadata

Architectural/technical:
- Reserved-namespace publish check now tests the resolved path component,
  host-agnostic, closing the absolute-topic subscription-event forgery gap.
- 401 only when no credential is presented; 403 for every presented-credential
  failure, so the status code no longer discloses validation vs scope.
- Matcher query-param names: 400 if malformed, 501 if well-formed but
  unimplemented, matching the claim-list handling.
- Define URITemplate matching (reverse-expansion) and its DoS limit.
- Specify full-string anchoring for Regexp matching.
- Subscriptions are created for registered and hub-specific matchers alike.
- State that payload claim-matching governs selection only, never authorization.

References:
- Add RFC 8174 (BCP 14 boilerplate).
- RFC 5988 -> 8288, RFC 4122 -> 9562, RFC 6982 -> 7942.
- Demote EventSource interface and XHR to informative.
- Convert inline DID links to a proper reference.

Editorial:
- Add Introduction; define canonical/alternate topic in Terminology.
- Clarify the JWS payload is a JWT.
- Registrable domain (eTLD+1) instead of "second-level domain".
- Rewrite Security Considerations as analysis plus pointers, no duplicated
  normative keywords; move the custom-matcher operator-review requirement
  into the body.
- Recommend 200 on publish, 201 NOT RECOMMENDED.

Metadata:
- Add ipr=trust200902 and abbrev; area -> Web and Internet Transport.

dunglas committed:
spec: cut matcher types to Exact + URLPattern, drop the registry

The pluggable matcher system (Regexp, CEL, URITemplate, the IANA matcher-type
registry, and reverse-DNS custom types) was speculative and unshipped. It
expanded the interoperability surface without a driving use case: a conforming
hub could share almost nothing with another beyond Exact, and the advanced
types had no second implementation.

Keep two mandatory types, Exact and URLPattern. URLPattern supersedes URI
Template, which lacked defined matching semantics (RFC 6570 specifies expansion,
not matching). Restore a reserved `*` wildcard for match-all.

- Subscription: only match/matchExact/matchURLPattern; unknown names -> 400.
- Topic matcher list: matchType MUST be Exact or URLPattern; drop the 501
  unsupported-type paths (both types are now mandatory).
- Remove the CEL sandbox and custom-matcher security subsections; fold the DoS
  note into URL-Pattern only.
- Remove the "Mercure Matcher Types" IANA registry and the cel/RFC 9485
  references.

RFC 6570 is retained for subscription-event URI Template expansion.

dunglas commented:
Superseded by https://github.com/dunglas/mercure/pull/1262, which carries the full spec rework: the matcher surface is reduced to two standards-based types (Exact + WHATWG URL Pattern), and authorization moves from the bespoke `mercure` claim to an OAuth 2.0 model (JWT access tokens RFC 9068, Rich Authorization Requests RFC 9396, Bearer usage RFC 6750, Protected Resource Metadata RFC 9728). Continuing there.
